src/dimension/Dimension.js

```diff
@@ -0,0 +1,488 @@
+            }
+            else {
+                this._dimContainsArray[i] = false
+            }
+        })
+    }
+    _initGroup(dataConnector) {
+        // tbd
+    }
+    // todo - maybe this or some other technique...?
+    _addPublicAPI(crossfilter) {
+        this.writeTopBottomQuery    = writeTopBottomQuery
+        // this.writeFilter = (queryBinParams) => writeGroupFilter(queryBinParams, this)
+        this.top                    = (k, offset, renderSpec, callback) => top(this, k, offset, renderSpec, callback)
+        this.writeTopQuery          = (k, offset, isRender) => writeTopQuery(this, k, offset, isRender)
+        this.writeTopBottomQuery    = (k, offset, ascDescExpr, isRender) => writeTopBottomQuery(this, k, ascDescExpr, isRender)
+        this.bottom                 = (k, offset, renderSpec, callback) => bottom(this, k, offset, renderSpec, callback)
+        this.writeBottomQuery       = (k, offset, isRender) => writeBottomQuery(this, k, offset, isRender)
+        // todo - temporary hack to support backwards compatibility
+        this.remove = this.dispose  = () => this.getCrossfilter().removeDimension(this)
+        this.value                  = () => this._dimArray
+        this.getCrossfilterId       = () => crossfilter.getId()
+        this.getProjectOn           = () => this._projectExpressions
+        this.setDrillDownFilter     = (value) => {
+            this._drillDownFilter = value
+            return this
+        }
+        this.getTable               = crossfilter.getTable
+        // testing haxx
+        this.getSamplingRatio       = () => this._samplingRatio
+    }
+    set(fn) {
+        this._dimArray = fn(this._dimArray)
+        return this
+    }
+    /******************************************************************
+     * private methods
+     */
+    /******************************************************************
+     * public methods
+     */
+    // todo - tricky: look at immerse/src/services/crossfilter.getTopN
+    group(groupId = false) { // todo - shouldn't this at least be parameterized like thus?
+        // if(!this._dimensionGroups.length) {
+        //     const newGroup = new Group(this.getDataConnector(), this)
+        //     return this.addGroupToDimension(newGroup)
+        // }
+        // else {
+        //     // todo - such an excellent question, eh? (tisws)
+        // }
+        const newGroup = new Group(this.getDataConnector(), this)
+        return this.addGroupToDimension(newGroup)
+    }
+    addGroupToDimension(newGroup) {
+        this._dimensionGroups.push(newGroup)
+        return newGroup
+    }
+    groupAll() {
+        return this.getCrossfilter().groupAll
+    }
+    setEliminateNull(eliminateNull) {
+        this._eliminateNull = eliminateNull
+        return this
+    }
+    getEliminateNull() {
+        return this._eliminateNull
+    }
+    /**
+     *  tbd public or private methods
+     */
```

Comment:
as in it's unclear if anything uses this method externally? we should commit to one or the other.